In [1]:
import itertools
import numpy as np

The Red Dragon Inn is an online Dungeons and Dragons community, connecting players worldwide. In this lab, you will develop a tool that helps users calculate the probabilities of specific events occurring during gameplay, enhancing their strategic planning. For instance, users may wish to calculate the probability of the sum of two six-sided dice being 10 or more. In Dungeons and Dragons, the notation mdn denotes the process of rolling m dice, each with n sides. Here, rolling two six-sided dice is expressed as 2d6, and rolling three eight-sided dice is denoted by 3d8. You are tasked with creating a utility to assist players in the following two scenarios:

A Calculating the probability of rolling m dice, each with n sides, and having the sum of the dice be greater than x. This calculation should only be implemented for m = 1 through 4 and n is at most 20.


In [2]:
def calculate_probability(m, n, x):
    """Calculates the probability of rolling m dice, each with n sides, and having the sum of the dice be greater than x.

    Args:
        m: The number of dice to roll (1-4).
        n: The number of sides on each die (1-20).
        x: The target sum.

    Returns:
        The probability of rolling a sum greater than x.
    """

    if not (1 <= m <= 4 and 1 <= n <= 20):
        raise ValueError("Invalid parameters: m and n must be between 1 and 20.")


    rolls = list(itertools.product(range(1, n+1), repeat=m))
    sums = np.array([sum(roll) for roll in rolls])
    probability = np.mean(sums > x)
    return probability



The first problem asks for the probability of the sum of the dice being greater than x, while the second problem asks for the probability of at least one die being greater than or equal to x.

In [3]:
def calculate_probability_2(m, n, x):
    """Calculates the probability of rolling at least one die greater than or equal to x.

    Args:
        m: The number of dice to roll.
        n: The number of sides on each die.
        x: The target value.

    Returns:
        The probability of at least one die being greater than or equal to x.
    """

    # Probability of a single die *not* being greater than or equal to x
    p_not_greater = (x - 1) / n

    # Probability of *all* dice *not* being greater than or equal to x
    p_all_not_greater = p_not_greater ** m

    # Probability of at least one die being greater than or equal to x
    probability = 1 - p_all_not_greater

    return probability



In [4]:
import itertools
# List out all possible rolls, in this case for 3d6
# rolls = list(itertools.product(range(1, 7), repeat=3))

# Create a list of all possible outcomes for rolling 3 six-sided dice (3d6).
rolls = list(itertools.product(range(1, 7), repeat=3))

# Example: Print the first 5 rolls
print(rolls[:5])


[(1, 1, 1), (1, 1, 2), (1, 1, 3), (1, 1, 4), (1, 1, 5)]


In [5]:
m = 2   # Number of dice
n = 6 # Number of sides on each die
x = 7 # Target value
probability = calculate_probability(m, n, x)
print("Probability:", probability)

Probability: 0.4166666666666667


In [6]:
m = 1   # Number of dice
n = 20 # Number of sides on each die
x = 17 # Target value
probability = calculate_probability(m, n, x)
print("Probability:", probability)

Probability: 0.15


In this case the probabilty of rolling 1d20 dice and getting a result of 17 or more is 15%.

In [7]:
m = 3  # Number of dice
n = 4  # Number of sides on each die
x = 10  # Target value
probability = calculate_probability(m, n, x)
print("Probability:", probability)

Probability: 0.0625


In this case the probabilty of rolling 3d4 dice and getting a result of 10 or more is 6%.

In [8]:
m = 4  # Number of dice
n = 6  # Number of sides on each die
x = 6 # Target value
probability = calculate_probability_2(m, n, x)
print("Probability:", probability)

Probability: 0.5177469135802468


Lastly: In this case the probabilty of rolling 4d6 dice and getting at least one dice to show a 6 is about 52%

References:

**Single Die Probability:** First calculated the probability of one die not being greater than or equal to x. This is the number of outcomes that are less than x divided by the total number of possible outcomes.

**Multiple Dice Probability:** Then calculated the probability of all dice not being greater than or equal to x. The rolls of each individual dice are independent events since one roll does not impact the chances of another roll. This allows us to multiply the probabilities of each die not being greater than x (the single die probability).

**Complement Rule:** Finally, we used the complement rule to find the probability of at least one die being greater than or equal to x. This is 1 minus the probability of all dice not being greater than or equal to x.Referncing the idea of P(A) + P(not A) = 1. Which was discussed in class.